clash_royale_service.py

```diff
@@ -168,4 +178,74 @@ def past_weeks_clan_war(self, clan_players_war_history: Dict[str, PlayerActivity
             player.war_active = compute_war_active(player.fame_hist)
             player.elder_worthy = compute_elder_worthy(player.fame_hist)
             player.avg_fame = compute_average_fame(player.fame_hist)
-        return clan_players_war_history
\ No newline at end of file
+        return clan_players_war_history
+
+    '''
+    Remaining war attacks
+    '''
+
+    def _get_participated(self, participants: List[Any]) -> int:
+        count = 0
+        for person in participants:
+            if person["decksUsedToday"] != 0:
+                count += 1
+        return count
+
+    def _get_decks_remaining(self, participants: List[Any], clan_members: List[Any]) -> int:
+        tag_to_player_war = {person["tag"]: person for person in participants}
+        tag_to_player_clan = [person["tag"] for person in clan_members]
+        num_players_left_clan = 0
+        num_decks_players_in_clan = 0
+        for player in tag_to_player_war:
+            if player not in tag_to_player_clan:
+                # Assuming that player leaving the clan will not come back, and they will take up all 4 decks
+                num_players_left_clan += 1
+            else:
+                num_decks_players_in_clan += tag_to_player_war[player]["decksUsedToday"]
+
+        return  MAX_DECK_PER_PLAYER*(PARTICIPANTS_LIMIT - num_players_left_clan) - num_decks_players_in_clan
+
+    def _get_players_remaining(self, participants: List[Any], clan_members: List[Any]) -> int:
+        # Identifies players in the clan that hasn't done their wars yet
+        tag_to_player_war = {person["tag"]: person for person in participants}
+        tag_to_player_clan = [person["tag"] for person in clan_members]
+
+        count = 0
+        for tag in tag_to_player_clan:
+            # Assumption: anyone in clan should also be in war_log_history
+            if tag not in tag_to_player_war:
+                log.info(
+                    f"player tag: {tag}, is not in currentriverrace response")
+                continue
+            if tag_to_player_war[tag]["decksUsedToday"] != 4:
+                count += 1
+        return count
+
+    def clan_remaining_war_attacks(self, clan_tag: str) -> List[ClanRemainingWarAttacks]:
+        curr_river_race = self.clash_royale_client.get_current_river_race(
+            clan_tag)
+
+        all_clan_attacks = []
+        for clan_river_info in curr_river_race["clans"]:
+            tag = clan_river_info["tag"]
+            # TODO: minor optimisation is could pull this out into separate for loop and make parallel API calls

```

Comment:
During your testing btw, how long is the overall discord command taking? Maybe we should do this optimisation (or can do in follow up PR after entire feature is done)

Reply:
not too slow. only 6 api calls!

Reply:
Can you time it?